### ISSUE TITLE
Using "const" widgets for "const" constructors, even with only single "Key" argument constructor. 

### ISSUE THREAD

**Issue by AniketMalik** [Milad-Akarie/auto_route_library#135]
This is more of an optimization request than an issue.

Version:
auto_route: ^0.5.0

Recently I discovered some benefits of using `const` constructors which helps in performance optimization for both stateless and statefull widgets while rebuilding the widgets.

Specific line from API docs:

> Use const widgets where possible. (This is equivalent to caching a widget and re-using it.)

<br>
Read more here:

[StatelessWidget - Performance Optimizations](https://api.flutter.dev/flutter/widgets/StatelessWidget-class.html)
[StatefulWidget - Performance Optimizations](https://api.flutter.dev/flutter/widgets/StatefulWidget-class.html)

### Suggestion

We could use a `const` widget where `const` constructors are defined even with a single `Key` argument. As keys are not required to pass explicitly, it can be ignored altogether. Flutter handles keys for `const` widgets automatically.

#### Example:

Suppose we have a `Profile` class with a `const` constructor like so:

```dart
class Profile extends StatelessWidget {
  const Profile({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const SomeWidget();
  }
}
```

This `Profile` class can be instantiated with a `const` keyword, not creating a `new` instance of the class every time the tree is rebuilt. We do not even need to pass the `Key` argument to the constructor as it is handled internally.

```dart
class MyApp extends StatelessWidget {
  const MyApp({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const Profile();
  }
}
```

I looked at the generated code for the classes defined with this type of `const` constructor which has only a `Key` argument. The generated code creates a separate `Argument` class with only this `Key` argument which is not actually required while using `const` constructors.

```dart
//Profile arguments holder class
class ProfileArguments {
  final Key key;
  ProfileArguments({this.key});
}
```

This creates the following code for the `Route Settings` for the case `Profile` which will not allow to create a `const` widget even when the Profile class is a `const` widget with a `const` constructor in itself as the `key` argument passed to the `Profile` is not a constant value. This will create a `New` instance of the `Profile` widget ( analyzed by dart analyzer ) every time a route action for `Profile` widget is performed.

```dart
case Routes.profile:
  if (hasInvalidArgs<ProfileArguments>(args)) {
    return misTypedArgsRoute<ProfileArguments>(args);
  }
  final typedArgs = args as ProfileArguments ?? ProfileArguments();
  return CupertinoPageRoute<dynamic>(
    builder: (context) => Profile(key: typedArgs.key),
    settings: settings,
  );
```

This can be prevented by totally excluding the need to create a separate `Argument` class only for the `Key` type argument by default. Creating a separate argument class even when there is only this `Key` type argument can be done using an `Annotation` type which will create the supporting `Argument` class allowing to pass a `unique key` argument when/where required. By excluding the `Argument` class in favor of the `const` widget instantiation the generated code can be as follows:

```dart
case Routes.profile:
  return CupertinoPageRoute<dynamic>(
    builder: (context) => const Profile(),
    settings: settings,
  );
```

This will also help in returning the cached version of the `Profile` widget without creating a `New` instance every time a route action for `Profile` widget is performed.

This additional `Argument` class is useful when there are __additional__ arguments __other__ than `Key` type. For example a `username` argument for the `Profile` class. Then we cannot create a `const` widget as the widget depends on a different value which cannot be determined at compile time.

```dart
//Profile arguments holder class
class ProfileArguments {
  final Key key;
  final String username;
  ProfileArguments({this.key, this.username});
}
```

I suppose that this minor change to the API can impact the performance positively as described in the official API docs.

**Comment by Milad-Akarie** [comment]
Hey @AniketMalik  That's great. I'm total into this. I'm recreating the base structor of auto_route these days and I'll get to this until I'm done with the basics.

**Comment by AniketMalik** [comment]
@Milad-Akarie that's great. Looking forward to this update. Cheers. Stay safe.

**Comment by Milad-Akarie** [comment]
Hey @AniketMalik  it's been implemented in auto_route: 0.6.4
There will not be a generated args class if a page constructor is marked as const and it has a single `Key` parameter.
also a const keyword is added before the page constructor in the route builder.
Thanks for your contribution.

**Comment by AniketMalik** [comment]
That's great news. Thank you. Cheers.
